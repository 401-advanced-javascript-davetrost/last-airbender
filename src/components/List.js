import React, { Component } from 'react';
import { getCharacters } from '../services/getCharacters';
import PropTypes from 'prop-types';
import { CharacterListDisplay } from './CharacterListDisplay';

export default class List extends Component  {

  state = {
    items: [],
    search: this.props.match.params.search || '*',
    page: this.props.match.params.page || 1,
  }

  apiGetCharacterList = () => {
    const { search, page } = this.state;
    const { perPage } = this.props;
    getCharacters({ search, page, perPage })
      .then(items => this.setState({ items }));
  }
  
  handlePageChange = increment => {
    const { search, page, items } = this.state;
    const { perPage, history } = this.props;
    if(items.length >= perPage) {
      const newPage = Math.max(1, +page + increment);
      this.setState({ page: newPage });
      history.push(`/list/${search}/${newPage}`);
    }
  }
  
  componentDidMount = () => {
    this.apiGetCharacterList();
  }

  componentDidUpdate = (previousProps) => {
    const { page: prevPage } = previousProps.match.params;
    const { page } = this.props.match.params;
    if(prevPage != page) this.apiGetCharacterList();
  }

  handleCharacterSelect = id => {
    const { history } = this.props;
    history.push(`/detail/${id}`);
  }

  render() {
    const items = this.state.items.map(item => (
      <CharacterListDisplay
        key={item._id || item.name}
        name={item.name}
        image={item.photoUrl}
        handleCharacterSelect={() => this.handleCharacterSelect(item._id)} />
    ));
  
    return (
      <>
        <button onClick={() => this.handlePageChange(-1)}>Prev</button>
        <button onClick={() => this.handlePageChange(+1)}>Next</button>
        <div>
          {items}
        </div>
      </>
    );
  }

  static propTypes = {
    match: PropTypes.shape({
      params: PropTypes.shape({
        search: PropTypes.string,
        page: PropTypes.string
      }).isRequired
    }).isRequired,
    history: PropTypes.object.isRequired,
    perPage: PropTypes.number.isRequired
  }

}
