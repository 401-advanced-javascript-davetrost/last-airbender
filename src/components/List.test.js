import React from 'react';
import { shallow } from 'enzyme';
import List from './List';

jest.mock('../services/getCharacters', () => ({
  getCharacters() {
    return Promise.resolve([
      {
        _id: '5cf5679a915ecad153ab6927',
        allies: ['Her husband'],
        enemies: ['Amon'],
        photoUrl: 'https://vignette.wikia.nocookie.net/avatar/images/f/f9/Fire_Nation_councilwoman.png/revision/latest?cb=20121110144847',
        name: 'Fire Nation councilwoman',
        gender: 'Female',
        hair: 'Gray',
        love: 'Her husband',
        weapon: 'Fire',
        profession: 'Councilwoman ',
        position: 'Fire Nation representative to the United Republic Council (formerly) Upper-class United Republic citizen',
        affiliation: 'Fire Nation United Republic Council (formerly)',
      }
    ]);
  }
}));

describe('List component', () => {
  it('renders', () => {
    const wrapper = shallow(<List 
      match={{
        params: {
          search: '',
          page: '1'
        }
      }}
      history={{}} 
      perPage={10}
    />);
    expect(wrapper).toMatchSnapshot();
  });
});
  
