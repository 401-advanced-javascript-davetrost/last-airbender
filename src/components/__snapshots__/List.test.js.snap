// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List component renders 1`] = `
<Fragment>
  <button
    onClick={[Function]}
  >
    Prev
  </button>
  <button
    onClick={[Function]}
  >
    Next
  </button>
  <div />
</Fragment>
`;
